### PR TITLE
Fix sscache on windows and make setup.py nicer

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: tools/setup.py [hunk 1]
```diff
@@ -44,29 +45,68 @@ def write_lastchange():
     # ])
 
 
-def get_gn_args():
+# If this text is found in args.gn, we assume it hasn't been hand edited.
+gn_args_header = [
+    "# This file is automatically generated by tools/build.py.",
```
Comment (ry) [inline tools/setup.py]: setup.py

File: tools/setup.py [hunk 2]
```diff
@@ -44,29 +45,68 @@ def write_lastchange():
...
 
     # Check if ccache or sccache are in the path, and if so we set cc_wrapper.
     cc_wrapper = find_executable("ccache") or find_executable("sccache")
     if cc_wrapper:
-        out += [r'cc_wrapper="%s"' % cc_wrapper]
-        # Windows needs a custom toolchain for cc_wrapper to work.
+        out += ['cc_wrapper="%s"' % cc_wrapper]
+        # For cc_wrapper to work on Windows, we need need a to select our own
+        # toolchain by overriding 'custom_toolchain' and 'host_toolchain'.
+        # TODO: Is there a way to use it without the involvement of args.gn?
```
Comment (ry) [inline tools/setup.py]: Any theories as to why it has to be done in args.gn ?

File: tools/util.py [hunk 3]
```diff
@@ -173,3 +173,32 @@ def parse_exit_code(s):
...
+
+    home = os.path.expanduser("~")
+    candidates = [
+        rel(path, base, "", ".") if base is not None else None,
+        rel(path, home, "~" + os.sep, "~"),
+        os.path.abspath(path)
+    ]
+    candidates = filter(None, candidates)
+    candidates.sort(key=len)
+    return candidates[0]  # Return the shortest.
```
Comment (ry) [inline tools/util.py]: Can you add a test for this to tools/util_test.py ?

File: tools/setup.py [hunk 4]
```diff
@@ -44,29 +45,68 @@ def write_lastchange():
...
+    assert gn_args_are_generated(lines)  # With header -> generated.
+
+    # Ensure the directory where args.gn goes exists.
+    dir = os.path.dirname(args_filename)
+    if not os.path.isdir(dir):
+        os.makedirs(dir)
+
+    with open(args_filename, "w") as f:
+        f.writelines("\n".join(lines) + "\n")
+
```
Comment (ry) [inline tools/setup.py]: I feel like this gn args writing is sufficiently complex as to warrant a unit test. Consider moving these functions to util.py and making a test in util_test.py

File: tools/util.py (hunk 3 above)
Comment (piscisaureus) [reply tools/util.py]: It seemed like a nice idea but I feel that it's too much code. I'm gonna remove it.

File: tools/setup.py (hunk 2 above)
Comment (piscisaureus) [reply tools/setup.py]: Well I can't set default_toolchain elsewhere. 
Except in .gn in the root but the current_os, is_win etc variables aren't available there.
So where else would it go?

File: tools/setup.py (hunk 1 above)
Comment (piscisaureus) [reply tools/setup.py]: Fixed.

File: tools/setup.py (hunk 4 above)
Comment (piscisaureus) [reply tools/setup.py]: Done.

**Comment by piscisaureus** [comment]
@ry PTAL

**Pull request merged**